File: midonet-client/src/main/java/org/midonet/client/resource/DhcpSubnet.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.client.resource;

import java.net.URI;
import java.util.List;

import org.midonet.client.VendorMediaType;
import org.midonet.client.WebResource;
import org.midonet.client.dto.DtoDhcpHost;
import org.midonet.client.dto.DtoDhcpOption121;
import org.midonet.client.dto.DtoDhcpSubnet;

public class DhcpSubnet extends ResourceBase<DhcpSubnet, DtoDhcpSubnet> {

    public DhcpSubnet(WebResource resource, URI uriForCreation, DtoDhcpSubnet
        principalDto) {
        super(resource, uriForCreation, principalDto, VendorMediaType
            .APPLICATION_DHCP_SUBNET_JSON);
    }

    /**
     * Gets URI for this DhcpSubnet
     *
     * @return
     */
    @Override
    public URI getUri() {
        return principalDto.getUri();
    }

    /**
     * Gets prefix for this subnet.
     *
     * @return
     */
    public String getSubnetPrefix() {
        return principalDto.getSubnetPrefix();
    }

    /**
     * Gets default gateway for this subnet.
     *
     * @return
     */
    public String getDefaultGateway() {
        return principalDto.getDefaultGateway();
    }

    /**
     * Gets opt121Routes.
     *
     * @return
     */
    public List<DtoDhcpOption121> getOpt121Routes() {
        return principalDto.getOpt121Routes();
    }

    /**
     * Gets DHCP server IP.
     *
     * @return
     */
    public String getServerAddr() {
        return principalDto.getServerAddr();
    }

    /**
     * Gets DNS server IP.
     *
     * @return
     */
    public List<String> getDnsServerAddrs() {
        return principalDto.getDnsServerAddrs();
    }

    /**
     * Gets interface MTU.
     *
     * @return
     */
    public int getInterfaceMTU() {
        return principalDto.getInterfaceMTU();
    }

    /**
     * Gets length of the subnet address.
     *
     * @return
     */
    public int getSubnetLength() {
        return principalDto.getSubnetLength();
    }

    /**
     * Sets default gateway.
     *
     * @param defaultGateway
     * @return this
     */
    public DhcpSubnet defaultGateway(String defaultGateway) {
        principalDto.setDefaultGateway(defaultGateway);
        return this;
    }

    /**
     * Sets host.
     *
     * @param hosts
     * @return this
     */
    public DhcpSubnet hosts(URI hosts) {
        principalDto.setHosts(hosts);
        return this;
    }

    /**
     * Sets opt121Routes.
     *
     * @param opt121Routes
     * @return this
     */
    public DhcpSubnet opt121Routes(List<DtoDhcpOption121> opt121Routes) {
        principalDto.setOpt121Routes(opt121Routes);
        return this;
    }

    /**
     * Sets SubnetLength.
     *
     * @param subnetLength
     * @return this
     */
    public DhcpSubnet subnetLength(int subnetLength) {
        principalDto.setSubnetLength(subnetLength);
        return this;
    }

    /**
     * Sets prefix of the subnet.
     *
     * @param subnetPrefix
     * @return this.
     */
    public DhcpSubnet subnetPrefix(String subnetPrefix) {
        principalDto.setSubnetPrefix(subnetPrefix);
        return this;
    }

    /**
     * Sets DHCP Server IP.
     *
     * @param serverAddr
     * @return this
     */
    public DhcpSubnet serverAddr(String serverAddr) {
        principalDto.setServerAddr(serverAddr);
        return this;
    }

    /**
     * Sets DNS Server IP.
     *
     * @param dnsServerAddr
     * @return this
     */
    public DhcpSubnet dnsServerAddrs(List<String> dnsServerAddr) {
        principalDto.setDnsServerAddrs(dnsServerAddr);
        return this;
    }

    /**
     * Sets Interface MTU.
     *
     * @param interfaceMTU
     * @return this
     */
    public DhcpSubnet interfaceMTU(int interfaceMTU) {
        principalDto.setInterfaceMTU(interfaceMTU);
        return this;
    }

    /**
     * Gets host resources under this subnet.
     *
     * @return
     */
    public ResourceCollection<DhcpHost> getDhcpHosts() {
        return getChildResources(
            principalDto.getHosts(),
            null,
            VendorMediaType.APPLICATION_DHCP_HOST_COLLECTION_JSON,
            DhcpHost.class,
            DtoDhcpHost.class);
    }

    /**
     * Adds subnet host resource under this subnet.
     *
     * @return new DhcpHost()
     */
    public DhcpHost addDhcpHost() {
        return new DhcpHost(resource, principalDto.getHosts(),
                              new DtoDhcpHost());
    }

    @Override
    public String toString() {
        return String.format("{DhcpSubnet, networkp=%s, length=%s}",
                             principalDto.getSubnetPrefix(),
                             principalDto.getSubnetLength());
    }
}

File: python-midonetclient/src/midonetclient/dhcp_subnet.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2013 Midokura PTE LTD.
# All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.


from midonetclient import dhcp_host
from midonetclient import resource_base
from midonetclient import vendor_media_type


class DhcpSubnet(resource_base.ResourceBase):

    media_type = vendor_media_type.APPLICATION_DHCP_SUBNET_JSON

    def __init__(self, uri, dto, auth):
        super(DhcpSubnet, self).__init__(uri, dto, auth)

    def get_default_gateway(self):
        return self.dto['defaultGateway']

    def get_server_addr(self):
        return self.dto['serverAddr']

    def get_dns_server_addrs(self):
        return self.dto['dnsServerAddrs']

    def get_subnet_prefix(self):
        return self.dto['subnetPrefix']

    def get_subnet_length(self):
        return self.dto['subnetLength']

    def get_interface_mtu(self):
        return self.dto['interfaceMTU']

    def get_opt121_routes(self):
        return self.dto['opt121Routes']

    def is_enabled(self):
        return self.dto["enabled"]

    def default_gateway(self, gw):
        self.dto['defaultGateway'] = gw
        return self

    def server_addr(self, addr):
        self.dto['serverAddr'] = addr
        return self

    def dns_server_addrs(self, addrs):
        self.dto['dnsServerAddrs'] = addrs
        return self

    def subnet_prefix(self, prefix):
        self.dto['subnetPrefix'] = prefix
        return self

    def subnet_length(self, length):
        self.dto['subnetLength'] = length
        return self

    def interface_mtu(self, mtu):
        self.dto['interfaceMTU'] = mtu
        return self

    def opt121_routes(self, routes):
        self.dto['opt121Routes'] = routes
        return self

    def enabled(self, enabled):
        self.dto['enabled'] = enabled
        return self

    def get_dhcp_hosts(self):
        query = {}
        headers = {'Accept':
                   vendor_media_type.APPLICATION_DHCP_HOST_COLLECTION_JSON}
        return self.get_children(self.dto['hosts'], query, headers,
                                 dhcp_host.DhcpHost)

    def get_dhcp_host(self, ip_addr, mac):
        """
        Given the IP address and mac, returns DhcpHost object
        """
        for dh in self.get_dhcp_hosts():
            if dh.get_ip_addr() == ip_addr and dh.get_mac_addr() == mac:
                return dh
        return None

    def add_dhcp_host(self):
        return dhcp_host.DhcpHost(self.dto['hosts'], {}, self.auth)
